rfcs/testdriver_bidi.md

```diff
@@ -0,0 +1,277 @@
+
+The specific protocol implementation would be required to opt-in for this protocol. Now it’s `EdgeChromiumDriverProtocol` and `ChromeDriverProtocol`.
+
+###### Dependency on `WebDriverProtocol`
+
+New [`WebDriverBidiProtocol`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR568) should be introduced. It is an extension of WebDriverProtocol, but with additional BiDi-specific protocol parts, e.g. [`BidiEventsProtocolPart`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-47192f72de48b834a50992ada793229686299b5165612c2d97af6811764514c7R325) and [`BidiScriptProtocolPart`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-47192f72de48b834a50992ada793229686299b5165612c2d97af6811764514c7R358). In order to enable the bidi session, the inherited `enable_bidi` flag should be overridden to `true`.
+
+###### Dedicated loop
+
+To be able to process events from the browser and from the tests in the same time, the WebDriverBidiProtocol protocol should have a dedicated [event loop](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR577), which will be used by wptrunner to run async actions and process events in WebDriver BiDi transport.
+
+###### Alternative
+
+[Add `AsyncProtocolPart` with `get_loop` method](#add-asyncprotocolpart-with-get_loop-method).
+
+##### Extend `TestExecutor`
+
+In this example we consider changes in [`WebDriverTestharnessExecutor.do_testharness`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dL591). Other `TestExecutor` (`MarionetteTestharnessExecutor`, `SeleniumTestharnessExecutor`, `ServoWebDriverTestharnessExecutor`,
+`WKTRTestharnessExecutor` etc) implementations should be done in the same way in order to support WebDriver BiDi.
+
+###### Messages from testdriver to wptrunner
+
+The communication channel between testdriver and wptrunner is set in the `do_testharness` method. Currently the messages from testdriver to wptrunner it is done via long poll by calling BaseProtocolPart.execute_async_script which is synchronous (async in the name indicates that the script is wrapped in a special way before sent to the browser), and blocks processing the event  loop in wptrunner, which in turn blocks the event processing while waiting for the next message form testdriver to wptrunner. To overcome this limitation, this RFC changes the wptrunner - testdriver communication channel to asynchronous non-blocking  [`BidiScriptProtocolPart.async_call_function`](https://github.com/web-platform-tests/wpt/pull/44649/files#diff-e5a8911dd97e0352b1b26d8ce6ef0a92b25378f7c9e79371a1eb1b1834bc9a8dR761) if available.  `WebDriverBidiProtocol`’s event loop should be used for message processing. This behavioral change is done only for protocols supporting BiDi.
```

Comment:
Is it `s/with the harness/with the testdriver`, or I misread the comment?
> ... but instead have the harness emit custom `script.message` events to communicate information **with the testdriver**...

Do you mean in case of BiDi, the testharness should still make a long poll for getting commands from testdriver, but emit `message` events directly to `window`, avoiding `__wptrunner_process_next_event`? I'm not sure if I'm getting the point.

Reply:
s/with the harness/with the browser/ yes.

In general the current way the testdriver event queue works, with `__process_next_event` is a workaround for limitations in WebDriver classic. I've been assuming that once we have BiDi we'd be able to move away from that model of polling an event queue in a single window, and instead use the browser event loop directly, and communicate via WebDriver BiDi events rather than blocking script evaluation. 

We don't necessarily have to do that right away, but I'd at least like to see what the migration path looks like,

Reply:
> Do you mean in case of BiDi, the testharness should still make a long poll for getting commands from testdriver, but emit `message` events directly to `window`, avoiding `__wptrunner_process_next_event`? I'm not sure if I'm getting the point.

Oh, I got it. You mean the harness can provide a `channel` for the client-side wptrunner. I guess it is possible and makes sense, but I'd prefer to consider it in a different RFC to reduce execution risks.

Reply:
If you have some time to prototype that I'd appreciate it. As mentioned one concern in general is that making the wrong first step could make it harder in the long run if we bake in assumptions that don't work for other use cases.

Reply:
Since the communication between the testharness and the browser is internal and it is not exposed to test authors, it looks like it can be addressed separately. What assumptions from this RFC could prevent us from changing the event implementation in the future? 

It sounds like a preload script + a channel for setting up biderectional communication would not require API changes on the testdriver.js side.